### ISSUE TITLE
Improving the DataBroker API

### ISSUE THREAD

**Issue by danielballan** [NSLS-II/dataportal#195]
Let's discuss how we might streamline the databroker API for common use cases. I think we can do this without breaking the existing API.

Having now seen many notebooks and IPython command histories in the wild, here are the problems I see with our first attempt at the design:
1. People are loading image data when all they needed was scalars (e.g., image stats) at great cost in performance.
2. The sequence of commands to get a simple table -- a DataFrame -- is too verbose.
3. The `StepScan` interface, meant to address (2), is too limiting. For example, there is no way to turn off filestore when all you wanted was some scalars.
4. The dual role of `db` as both a thing that gets headers and a module-like thing with functions is muddy. Personally I find it a little weird to explain, and it's not a pattern I have seen elsewhere. (As it happens, it's my own design, so I'm criticizing it with impunity.)

Here is my rough proposal to kick off discussion. I would give more weight to any opinions from @klauer, @cmazzoli, @ambarb, @afluerasu, @yugangzhang, @pavoljuhas, @ordirules and any others who have actually used this thing at a beamline.
### Load Data into DataFrame

Instead of `StepScan`, something like this would let us build in more flexibility.

```
from datapotal import DataBroker as db, get_events, make_df
header = db[123]
make_df(header)
```

This would also work on data from multiple scans.

```
headers = db[-10:]  # last ten scans
make_df(headers)
```

To be clear, that last line is an alias for this mouthful: `DataMuxer.from_events(get_events(header)).to_sparse_dataframe()`.

But, I hear you say, `StepScan` does the above in just one line! The payoff of separating header look-up from DataFrame construction is _much better performance_ in many cases. Behold:
### Load Select Fields into DataFrame

```
from datapotal import make_df
header = db[123]
make_df(header, ['det1', 'det2', 'motor3'])  # only loads filestore data as needed
```
### Load Images into a Lazy-Loading Iterable

I think the image-loading API we have is actually fine, but it needs to be documented and advertised. It is not widely used.

```
from dataportal import Images
header = db[123]
imgs = Images(header, 'det1')
```
### Search for headers when scan ids / uids are not known

Currently, we have `db.find_headers`, which is both verbose and somewhat odd, as I said in (4) above.

Instead, we might simply make `db` callable, so that `db.find_headers(owner='dallan')` becomes `db(owner='dallan')`.

In summary, `db` becomes an object that one does one thing: it looks up headers. You can slice `db[]` and call `db()`, and that's it. The methods `db.fetch_events` and `db.find_headers` can be retired, replaced by `get_events` (a top-level function) and `db()` respectively. I see no reason to actually _remove_ these things and break existing code, but we can stop documenting them.
### Let the Debate Commence!
1. Are there other important workflows not addressed here?
2. Is this an improved compromise between verbosity/flexibility? Can you think of a better one?
3. Let's save the inevitable debate about naming things for the end.

**Comment by yugangzhang** [comment]
```
I have one suggestion:

Some detectors, like CHX currently used Eiger, save h5 files containing lots of useful parameters, such as wavelength, detector distance, beam center, et.al.   We  added these parameters to eiger class and  updated eiger.py in 
https://github.com/yugangzhang/chxtools/blob/master/chxtools/pims_readers/eiger.py.
It would be nice to read all these info from the headers or your make_df. 

```

**Comment by tacaswell** [comment]
@yugangzhang All of that data _should_ be going into meta-data store so that it is accessible/searchable .

**Comment by yugangzhang** [comment]
I know. It will be convenient for users that we can incorporate these information into make_df or other functions.  But that (make_df) function should not read the image data.

**Comment by jrmlhermitte** [comment]
I am fine with any storage method so long as the tables describing the motors or other scalars may easily be associated with their img/raw data via some unique identifier. This would allow for a two-way access (i.e. if I have the img sequence, I can get the dataframes for the motors, or vice versa without much extra work.)

A two step process makes sense for the retrieval of the data. I think it is a good idea to force the user to understand the abstraction layers to some level without having to do more work than follow a manual and this would do so. Forcing two functions to retrieve the expt data and image data would help.

@tacaswell and @yugangzhang bring good points. I would like to add that I personally prefer that any parameter describing the expt should be kept as close to the source as possible rather than a searchable grab bag of all data in one header. I.e. it makes more sense to keep parameters such as an exposure time in the EIGER header file (which only matters for the EIGER) as opposed to some lower level dataframe describing the motors. So I would be more a proponent of a tree like structure of searchable bags.
Some other more fuzzy parameters would also also have to be decided on a case by case basis. For example the wavelength and beam center. The former probably should be saved in both db and eiger (because eiger will process data differently for different wavelengths, like wavelength dependent flatfield correction etc. Since eiger can independently change this parameter, it might accidentally not match the beamline). The latter should probably be saved in EIGER for the sake of portability and ease of self-contained data analysis.

So I'm basically saying:
1. Separation of reading different data sets should be forced upon the user by independent calls where it helps the user better understand the abstraction layers without much work.
2. I think some data should not be stored into the databroker, like exposure time for EIGER.

What do you think? You guys see things on a much larger scale than I am seeing so I am open to any changes.
Thanks, the eiger reader and the pims readers work great. :-)

**Comment by afluerasu** [comment]
These are some good points. I also realized that some quantities (what @ordirules 
calls 'fuzzy parameters') are actually not associated with any EPICS PV's and hence are not recorded in any searchable database. The beam center - X,Y coodinates on the Eiger - and detector distance are such an example. I'm not even sure of we have 'wavelength' or 'bandwidth' PVs at this moment ...
In any case, I think that the right philosophy is the one we were following so far, where all the metadata is in the metadata store. If I'm interested in knowing the detector distance, I may actually think about doing it in an automatic way (e.g. reading an encoder) and this will automatically result in the associated PV being available.

The problem is that accessing such metadata associated with e.g. images (or scalars) that we opened following some of the simplified ways described above by @danielballan is not necessarily very simple or clear. I think that as a part of this discussion focused on improving the databroker API and 'standardizing' the methods we use to access data, we should included also accessing the metadata.

**Comment by danielballan** [comment]
@afluerasu and @ordirules: Yes, the so-called "fuzzy parameters" that are not recorded on PVs should be recorded in metadatastore. This can be done in a persistent way (set a parameter once, and it gets saved on all future scans until it's updated) or on a scan-by-scan basis. We can talk about how to do this before the beam comes back. I agree with @ordirules point that we should structure where things are stored, not just put everything in a flat "bucket."

A related issue is providing nice access to channel archiver data. We have a way to do this in the works.

**Comment by danielballan** [comment]
Thanks for the feedback, everyone. Updates have been made. The new API is faster and simpler and close to what I presented here. I'll post a link when the documentation is ready (before the beam comes back).

**Comment by klauer** [comment]
Great - looking forward to these changes.
